content/300-guides/025-migrate/500-data-migration.mdx

```diff
@@ -0,0 +1,300 @@
+
+<cmdResult>
+
+<cmd>
+
+```bash no-copy
+Prisma schema loaded from prisma/schema.prisma
+Datasource "db": PostgreSQL database "data-migration", schema "public" at "localhost:5401"
+
+Error:
+⚠️ We found changes that cannot be executed:
+
+  • Step 1 Added the required column `status` to the `Post` table without a default value. There are 4 rows in this table, it is not possible to execute this step.
+
+You can use prisma migrate dev --create-only to create the migration file, and manually modify it to address the underlying issue(s).
+Then run prisma migrate dev to apply it and verify it works.
+```
+
+</cmd>
+</cmdResult>
+
+Exit from the migration step and update the schema by adding a default value for the `status` field by adding the `@default()` attribute function.
+
+```prisma file=prisma/schema.prisma highlight=13;edit
+model User {
+  id    Int     @id @default(autoincrement())
+  email String  @unique
+  name  String?
+  posts Post[]
+}
+
+model Post {
+  id        Int      @id @default(autoincrement())
+  title     String
+  content   String?
+  published Boolean? @default(false)
+  status    Status   @default(Unknown)
+  author    User?    @relation(fields: [authorId], references: [id])
+  authorId  Int?
+}
+
+enum Status {
+  Unknown
+  Draft
+  InProgress
+  InReview
+  Published
+}
+```
+
+Re-run `prisma migrate dev` to generate and apply a migration to the database
+
+## Create and run the data migration file
+
+### Create a data migration file
+
+<Admonition>
+
+**Checkpoint**
+
+1. Push your changes to a remote origin and create a new pull request.
+1. Once you’re happy with the changes, merge the changes to your `main` branch.
+
+To apply the changes to your production database, add `prisma migrate deploy` as part of your deployment/ build step in CI
+
+</Admonition>
+
+### Run the data migration
+
+Inside the generated migration folder from the previous step, create a file called `data-migration.ts` file. This file will contain a data migration which will be implemented using Prisma Client.
+
+Add the following code to migrate the data from the `published` field to the `status` field in the file you just created:
+
+Update your `package.json` file to include the data migration file:
+
+```ts file=prisma/migrations/20230417131956_add-status-column/data-migration.ts
+import { PrismaClient } from '@prisma/client'
+
+const prisma = new PrismaClient()
+
+async function main() {
+  const posts = await prisma.post.findMany()
+
+  await prisma.$transaction(
+    posts.map((post) =>
+      prisma.post.update({
+        where: {
+          id: post.id,
+        },
+        data: {
+          status: post.published ? 'Published' : 'Unknown',
+        },
+      })
+    )
+  )
+  console.log('Data migrated successfully')
+}
+```
+
+The data migration is wrapped in a transaction to ensure that the query is rolled back, allowing you to iterate on your data migration file
```

Comment:
would `updateMany` work here as well? prisma.io/docs/reference/api-reference/prisma-client-reference#updatemany

Reply:
Not quite, as the data of the existing records has to be fetched first to update existing records based on the value of an existing field.